### ISSUE TITLE
Angular (+ CLI) 6.0.0: Global styles for libraries

### ISSUE THREAD

**Issue by jhuenges** [angular/angular-cli#10869]
### Versions
```
Angular CLI: 6.0.1
Node: 8.9.4
OS: win32 x64
Angular:
...

Package                      Version
------------------------------------------------------
@angular-devkit/architect    0.6.1
@angular-devkit/core         0.6.1
@angular-devkit/schematics   0.6.1
@schematics/angular          0.6.1
@schematics/update           0.6.1
rxjs                         6.1.0
typescript                   2.7.2
```

### Repro steps

* Step 1: generate a new library
* Step 2: add some global styles to the library
* Step 3: run `ng build your-library`
* Step 4: global styles are not in `./dist`

### Observed behavior
The global styles from the library are not copied/used 

### Desired behavior
Global styles inside a library should be part of the generated output. I know that I could use component level styles (this works). However this doesnt work in all use cases. For example I have some css to style buttons but I dont have a `ButtonComponent` (and I dont want one) since I am only modifying the html `<button>` tag with the css and I am not adding some behaviour.

**Comment by kroeder** [comment]
I tried to add `styles: []` in my `angular.json` but get a 
```
Schema validation failed with the following errors:
  Data path "" should NOT have additional properties(styles).
```

https://github.com/angular/angular-cli/wiki/stories-global-styles shows, how to use global styles. This does not work when you add them into your library json configuration

```
  "my-library": {
      [...]
      "architect": {
        "build": {
          "builder": "@angular-devkit/build-ng-packagr:build",
          "options": {
            [...]
            "styles": [
              "styles/styles.scss"
            ]
  [...]
```

The documentation feels misleading since this does work for app builds but not for libraries.
Or am I doing something wrong?

**Comment by clydin** [comment]
The options are particular to the referenced builder (in this case, `"builder": "@angular-devkit/build-ng-packagr:build",`).  Both it and `ng-packagr` would need to add support to allow for such an option.

**Comment by kroeder** [comment]
Okay - I've sued ng-packgr before. Afaik there's no support for global css in ng-packagr and I had to add own build steps in order to get this running

Since having a global css file seems to me like a common thing, is there any plan to collab with the ng-packagr project on this?

**Comment by jhuenges** [comment]
Any update on this?

**Comment by kroeder** [comment]
@jhuenges if it helps, here's a gulp task snippet as workaround (or long-term solution, depending on what the angular team is planning 😃  )

**The snippet includes:**
- Building with `ng build my-lib --prod`
- Merging different (s)css files into one and concat them
- Copy additional assets, e.g. font-files 

```
const gulp = require('gulp');
const sass = require('gulp-sass');
const concat = require('gulp-concat');
const run = require('gulp-run');

const projectName = 'your-library';
const distDir = './dist/' + projectName;

gulp.task('build:angular-package', ['build:prepare'], function () {
    return run('ng build ' + projectName + ' --prod', {}).exec();
});

gulp.task('build:sass', ['build:angular-package'], function () {
    return gulp.src(['./src/lib/styles/styles.scss', './src/assets/**/*.scss'])
        .pipe(sass().on('error', sass.logError))
        .pipe(concat('styles.css'))
        .pipe(gulp.dest(distDir));
});

gulp.task('build:copy-assets', ['build:angular-package'], function () {
    return gulp.src(['src/assets/fonts/icons8/*', 'src/assets/fonts/opensans/v15/*'])
        .pipe(gulp.dest(distDir));
});

gulp.task('build', [
    'build:angular-package',
    'build:sass',
    'build:copy-assets'
]);
```

I had to change a couple of things since we have a slightly different custom-build but it should give an idea on how you can solve this.

Of course you need to install the necessary gulp-packages.

**Comment by jhuenges** [comment]
Thanks! I will try to use this, but a official solution would be great :)

**Comment by cm0s** [comment]
Here is another solution using only node package. In your package.json add dependency to package `cpx` and `scss-bundle`.
Then add the following entries in your package.json "scripts" property : 

```
"scripts": {
...
"build-mylib": "ng build mylib && npm run build-mylib-styles && npm run build-mylib-assets",
"build-mylib-styles": "cpx \"./projects/mylib/src/lib/style/**/*\" \"./dist/mylib/style\" && scss-bundle -e ./projects/mylib/src/lib/style/_other-styles.scss -d ./dist/mylib/style/_other-styles.scss",
"build-mylib-assets": "cpx \"./src/assets/**/*\" \"./dist/mylib/assets\"",
}
```

**Comment by gus-canya** [comment]
Hi. I wrote another solution that adapts to the `projects/**` directory structure for multiple library projects:

```
./projects
  my-lib/
    src/ 
      lib/
        my-lib-component/
           my-lib.component.css
./src
  assets/
    sass/
      projects/
       my-lib/
         my-lib-component/
           my-lib.component.scss
```

gulpfile.js: 

```
let gulp = require('gulp');
let sass = require('gulp-sass');
let autoprefixer = require('gulp-autoprefixer');
const aliases = require('gulp-style-aliases');

let fs                = require('fs');
let path              = require('path');
let cssTaskDictionary = [];
let cssTaskList       = [];
let watchTaskList     = [];

// SRC PATH definitions
let publicFolder = './projects';
let srcFolder = './src/assets';

let cssSrcPath = `${srcFolder}/sass/projects`;
let cssDest    = `${publicFolder}`;

// Gather Scss src files to watch and compile
(fs.readdirSync(cssSrcPath) || []).filter(directory => {
  let isDirectory = fs.lstatSync(path.join(cssSrcPath, directory)).isDirectory();
  return  !/global/.test(directory) &&
          !/theme/.test(directory) &&
          isDirectory;
}).forEach(project => {
  (fs.readdirSync(path.join(cssSrcPath, project)) || []).filter(projectCtrl => {
    return fs.lstatSync(path.join(cssSrcPath, project, projectCtrl)).isDirectory();
  }).forEach(directory => {
    fs.readdirSync(path.join(cssSrcPath, project, directory)).forEach(file => {
      cssTaskDictionary.push({ project: project, directory: directory, file: file });
    });
  });
});

cssTaskDictionary.forEach(taskDef => {

  let file = taskDef.file.replace(/\.scss/, '');
  file = file.replace(/_/, '');
  let taskSuffix = '-' + taskDef.project + '-' + taskDef.directory + '-' + file;
  let taskName = 'css' + taskSuffix;
  cssTaskList.push(taskName);

  // Sass will watch for changes in this file
  let srcPathFile = path.join(cssSrcPath, taskDef.project, taskDef.directory, taskDef.file);

  let destination = path.join(cssDest, taskDef.project, '/src/lib', taskDef.directory);

  gulp.task(taskName, () => {
    gulp.src([srcPathFile])
      .pipe(aliases({
        '@bootstrap': 'node_modules/bootstrap/scss',
        '@ebm': 'node_modules/ebm',
        '@sass': 'src/assets/sass',
        }))
      .pipe(sass({
        outputStyle: 'compressed'
      }).on('error', sass.logError))
      .pipe(autoprefixer({
        browsers: ['last 2 versions'],
        cascade: false,
        flexbox: true,
        }))
      .pipe(gulp.dest(destination)
    );
  });

  // Instantiate directory specific watch tasks
  let watchTaskName = 'watch-' + taskName;
  watchTaskList.push(watchTaskName);
  gulp.task(watchTaskName, () => {
    gulp.watch([srcPathFile], [taskName]);
  });
});

gulp.task('watch', watchTaskList);
```

**Comment by h-arora** [comment]
This might be a good read when a official solution is not yet ready: https://medium.com/@Dor3nz/compiling-css-in-new-angular-6-libraries-26f80274d8e5

**Comment by ramandeep-singh-1983** [comment]
> Here is another solution using only node package. In your package.json add dependency to package `cpx` and `scss-bundle`.
> Then add the following entries in your package.json "scripts" property :
> 
> ```
> "scripts": {
> ...
> "build-mylib": "ng build mylib && npm run build-mylib-styles && npm run build-mylib-assets",
> "build-mylib-styles": "cpx \"./projects/mylib/src/lib/style/**/*\" \"./dist/mylib/style\" && scss-bundle -e ./projects/mylib/src/lib/style/_style.scss -d ./dist/mylib/style/_styles.scss",
> "build-mylib-assets": "cpx \"./src/assets/**/*\" \"./dist/mylib/assets\"",
> }
> ```
> The build-mylib-styles works in two steps. First it copys all styles files found in the lib/style folder and then it uses the scss-bundle package to bundle all imports of the _style.scss file. I have other files in the lib/style folder without import so no need to bundle them.

Thanks! this solution works well. However, the new watch mode for the library doesn't work after this. So, one needs to rebuild and serve the page manually after any code change. Any way to make the watch mode work with this?

**Comment by amirofy2002** [comment]
but while packaging angular try to remove the script section of package.json and the following message occurs:

**Removing scripts section in package.json as it's considered a potential security vulnerability.**

**Comment by jhuenges** [comment]
Why would you do that?!

**Comment by mattiLeBlanc** [comment]
Would love to know when there is a general solution. For now I am just creating a `core-component` which loads in the styles which I have to reference once above my `app-root`. Not as nice, but works for now.

**Comment by cm0s** [comment]
> @ramandeep-singh-1983 i got it to work with watch mode, for everything except the scss-bundle, which i guess has to be a post-build step.
> 
> i used [concurrently](https://www.npmjs.com/package/concurrently) and [rimraf](https://www.npmjs.com/package/rimraf), in addition to the already mentioned [cpx](https://www.npmjs.com/package/cpx) and [scss-bundle](https://www.npmjs.com/package/scss-bundle)
> 
> here's the scripts i developed for it:
> 
> ```
>   "scripts": {
>     "clean": "rimraf \"./dist/*\"",
>     "scss-bundle": "scss-bundle -e \"./framework/src/styles/_framework.scss\" -d \"./dist/framework/styles/_framework.scss\"",
>     "watch-ts": "ng build --watch=true",
>     "watch-scss": "cpx \"./framework/src/styles/**/*\" \"./dist/framework/styles\" --watch --verbose",    
>     "watch-assets": "cpx \"./framework/src/assets/**/*\" \"./dist/framework/assets\"  --watch --verbose",
>     "watchAll": "npm run clean & concurrently \"npm:watch-*\""
>   },
> ```
> i put in the clean step because i have configured my ng-package.json with `"deleteDestPath": false`
> this allows me to control the deleting of the output folder myself.
> that way `ng build --watch` doesn't keep wiping out the work done by `cpx` every time a .ts file changes
> 
> i opened an issue in scss-bundle repo to ask them about adding a -watch mode
> [SimplrJS/scss-bundle#57](https://github.com/SimplrJS/scss-bundle/issues/57)

@hughanderson4 your solution is great, but I encountered a problem while using it. The `ng build --watch` command automatically clean the dist folder. When running the assets/css/bundle scripts in parallel with the ng build, the cpx and scss-bundle commands are executed before the ng build dist cleaning resulting in the deletation of the newly copied files.

To overcome this situation :
- Use Angular-cli 7.1 which has a new option (--deleteOutputPath=false) and use the rimraf command to enrue dist cleaning is proceeded before any other script is executed.
- If using previous Angular version : add a short delay before using cpx, scss-bundle (using npm package like @sindresorhus delay-cli)

**Comment by hughanderson4** [comment]
Yeah buddy similarly in my lib I got a ng-package.json and I had to modify like this to prevent the cleaning of the output folder:

```
{
  ...
  "deleteDestPath": false,
  ...
}
```

That why I manually put in the `rimraf` step.  Since `ng build` not doing that for me anymore.

I had problems trying to run all of it concurrently.  My workflow is to start the `watch-ts` and then after it has done the initial build then I start the `watch-assets` and `watch-scss` commands (if needed for that dev session), which is working pretty well for me.

**Comment by cm0s** [comment]
Great, thanks for sharing this solution. I didn't think about directly checking ng-packagr config options. Now I understand why you were using rimraf. It makes more sense ;).

I'm able to run the watchers together and everything seems to work fine. I use [npm-run-all]( https://github.com/mysticatea/npm-run-all/blob/master/README.md) to run parallel watch tasks, but it probably works the same way concurrently works (I didn't take the time to check).

**Comment by hughanderson4** [comment]
Great job, yeah npm-run-all is about the same as concurrently, I selected concurrently since it has wild card support to run bunch of jobs at same time.  ✌️

**Comment by ramandeep-singh-1983** [comment]
Thanks @hughanderson4, the watch mode is working nicely for me now with the following code:
`"scripts": {
    "clean": "rimraf \"./dist/*\"",
    "scss-bundle": "scss-bundle -e \"./projects/my-lib/src/lib/shared/styles/styles.scss\" -d \"./dist/my-lib/styles/styles.scss\"",
    "watch-ts": "ng build --watch=true",
    "watch-scss": "cpx \"./projects/my-lib/src/lib/shared/styles/**/*\" \"./dist/my-lib/styles\" --watch --verbose",    
    "watch-assets": "cpx \"./my-lib/src/assets/**/*\" \"./dist/my-lib/assets\" --watch --verbose",
    "watchAll": "npm run clean & concurrently \"npm:watch-*\""
  },`

However, i have 1 issue:
The dist folder has 2 sub-directories:
app, my-lib

the assets and styles are copied inside the my-lib directory after running "npm run watchAll". Now, how do i use the dist/my-lib\styles\_styles.scss in my application? I have made the application's styles.scss empty as i want all library specific global styles to go under my-lib/styles/_style.scss

**Comment by hughanderson4** [comment]
Hi @ramandeep-singh-1983 

> "scss-bundle" was not being called 

Yeah I was launching that separately back then.  Now, here is my latest package.json commands which work really nice.  I need to run only the `watch` command and it does everything:

```
  "scripts": {
    "clean": "rimraf \"./dist/*\"",
    "watch": "npm run clean && concurrently \"ng build --watch=true\" \"npm:watch-scss-*\" \"npm:watch-assets\"",
    "watch-scss-framework": "scss-bundle -e \"./framework/styles/_framework.scss\" -d \"./dist/framework/styles/_framework.scss\" -w \"./framework/styles\"",
    "watch-scss-variables": "scss-bundle -e \"./framework/styles/variables.scss\" -d \"./dist/framework/styles/variables.scss\" -w \"./framework/styles\"",
    "watch-assets": "cpx \"./framework/assets/**/*\" \"./dist/framework/assets\" --watch --verbose"
  }
```
I bundled my variables.scss file separately so users of my framework could refer to them without importing entire framework.scss.

To answer your question good sir, in your app (not lib) go to the angular.json and add this:

```
            "stylePreprocessorOptions": {
              "includePaths": [
                "/dist/my-lib/styles/"
              ]
            },
```

Ater that, in your app in any .scss file you should be able to write `@import "styles";` to bring in your lib's styles.scss.

**Comment by ramandeep-singh-1983** [comment]
Thanks! it worked. I also referred this article about global styles: https://scotch.io/tutorials/angular-shortcut-to-importing-styles-files-in-components

**Comment by ramandeep-singh-1983** [comment]
Another question:
My global style is being picked up from the library scss file now but what about the assets directory.
Right now, the dist directory has this structure:

my-app

app js files
assets
my-lib:

assets
styles
Suppose my library is a full fledged GUI library. The global styles reside in my library and are picked up by the app using @import statement. But what about the assets directory (icons, fonts etc.). They are part of my library but i have to copy them over to the my-app/assets directory for it to work. The my-lib directory resides outside the my-app directory in the dist directory so is it possible to pick up assets from the my-lib directory?

The my-app should be independent of my-lib so ideally it shouldn't need to contain assets for my-lib and all assets belonging to my-lib should be part of my-lib directory only.

**Comment by hughanderson4** [comment]
@ramandeep-singh-1983 My `watch-assets` command is just a `cpx` that gonna copy my assets from my lib project to my lib project's `dist\assets` folder.

Then I have to config each of my client apps to do this glob:
```
            "assets": [
              "src/favicon.png",
              "src/assets",
              {
                "glob": "**/*",
                "input": "node_modules/@dynatron/framework/assets/",
                "output": "./assets/framework"
              }
            ]
```
This goes in the angular.json under `projects\MyAwesomeApp\architect\build\options`
This glob gonna copy all my framework assets into my client site's `assets/framework` folder where they can be used.  That way my framework could provide multimedia assets to be used by all clients.

Hope it helps bro!  🍺

**Comment by mgechev** [comment]
Excellent discussion here! I can already see a few workarounds which seem reasonable. Although I'd discourage global styles in libraries, if your use case require such I'd recommend you to look at `ngx-build-plus`. You can customize your build using it.

**Comment by animeshrajurkar** [comment]
Thanks @hughanderson4 for your help, the assets thing worked too :)
@mgechev, thanks for the info about ngx-build-plus, i will explore it.

About the need to bundle global style in libraries: I am creating a UI application as a library and don't want to be dependent on the base app for the styles or assets, so my requirement is to have everything inside the library itself.

**Comment by ramandeep-singh-1983** [comment]
Thanks @hughanderson4 for your help, the assets thing worked too :)
@mgechev, thanks for the info about ngx-build-plus, i will explore it.

About the need to bundle global style in libraries: I am creating a UI application as a library and don't want to be dependent on the base app for the styles or assets, so my requirement is to have everything inside the library itself.

**Comment by vinagreti** [comment]
Hey guys.

The best solution IMHO is using `cpx` in the package.json to copy the css files and use `npm run build.lib` when building your library:

    "scripts": {
    ...
      "build.lib": "ng build my-lib && npm run copy-styles",
      "copy-styles": "cpx \"./projects/my-lib/src/lib/assets/styles/**/*.scss\" \"./dist/my-lib/assets/styles/\"",
    ...
    }

In your main application configure the tsconfig `PATHS` for pointing to your lib directory and than use your lib as you do with installed ones.

tsconfig.json

    {
      ...
      "compilerOptions": {
       ...
        "paths": {
          "my-lib": [
            "projects/my-lib"
          ],
         ...
        }
      }
    }

my-component.scss

    @import '~my-lib/assets/my-style';

When the cli notice the `~` in the import it will try to load from the configured path and if nothing is found it will look inside `node_modules` so you can use the local lib or the installed one without caring about changing the imports. It is like simulating an installation of the lib.

With this approach you can use paths everywhere to access different places in your cli app.

**Comment by ermarkar** [comment]
@ramandeep-singh-1983  Thanks for your answer.

1) But isn't `ng build --watch=true` will build the app itself rather than the library. 

2) And My library is dependent on materialize-css and I need to import its scss files into lib's global scss, but this is [not working](https://stackoverflow.com/questions/53985034/bundling-third-party-scss-files-into-angular-library-created-using-angular-cli)  until I installed the materialize-css to my application project. Is this the only way?

3) As I am using the scss-bundle and this will bundle all the scss files into one but there is an scss files that imports some resources like 

> @font-face {
>     font-family:"Roboto Regular";
>     font-weight:normal;
>     font-style:normal;
>     src: url("../../resources/fonts/roboto-regular/Roboto-Regular.eot?") format("eot"),  /* IE9 Compat Modes */
>          url("../../resources/fonts/roboto-regular/Roboto-Regular.woff") format("woff"), /* Pretty Modern Browsers */
>          url("../../resources/fonts/roboto-regular/Roboto-Regular.ttf") format("truetype");  /* Safari, Android, iOS */
>          /* url("../../resources/fonts/roboto-regular/Roboto-Regular.svg#Roboto-Regular") format("svg"); */ /* Legacy iOS */
> }

and I copied the resources folder as well but as there is only one bundled scss file for all scss (no folder hierarchy) so **../../resources** fails as there need only one **../resources**, what is the proposed solution ?

**Comment by ramandeep-singh-1983** [comment]
> @ramandeep-singh-1983 Thanks for your answer.
> 
> 1. But isn't `ng build --watch=true` will build the app itself rather than the library.
>> As i mentioned, i run 2 commands in parallel, 'npm run watch' watches and rebuilds the library and 'ng serve' watches and rebuilds the application.

> 2. And My library is dependent on materialize-css and I need to import its scss files into lib's global scss, but this is [not working](https://stackoverflow.com/questions/53985034/bundling-third-party-scss-files-into-angular-library-created-using-angular-cli)  until I installed the materialize-css to my application project. Is this the only way?
>> Looks like you were able to solve this by updating the library's ng-package.json 
> 3. As I am using the scss-bundle and this will bundle all the scss files into one but there is an scss files that imports some resources like
> 
> > @font-face {
> > font-family:"Roboto Regular";
> > font-weight:normal;
> > font-style:normal;
> > src: url("../../resources/fonts/roboto-regular/Roboto-Regular.eot?") format("eot"),  /* IE9 Compat Modes _/
> > url("../../resources/fonts/roboto-regular/Roboto-Regular.woff") format("woff"), /_ Pretty Modern Browsers _/
> > url("../../resources/fonts/roboto-regular/Roboto-Regular.ttf") format("truetype");  /_ Safari, Android, iOS _/
> > /_ url("../../resources/fonts/roboto-regular/Roboto-Regular.svg#Roboto-Regular") format("svg"); _/ /_ Legacy iOS */
> > }
> 
> and I copied the resources folder as well but as there is only one bundled scss file for all scss (no folder hierarchy) so **../../resources** fails as there need only one **../resources**, what is the proposed solution ?

>> I struggled with paths too but i was able to fix the issues eventually. Do you still have this issue?

**Comment by ermarkar** [comment]
this is how i done this https://stackoverflow.com/questions/54003034/scss-compilation-issue-in-angular-lib-created-using-angular-cli

**Comment by jbojcic1** [comment]
> Excellent discussion here! I can already see a few workarounds which seem reasonable. Although I'd discourage global styles in libraries, if your use case require such I'd recommend you to look at `ngx-build-plus`. You can customize your build using it.
> 
> Closing the issue here. If you think we're missing something and it's essential for the CLI to support this out of the box (i.e., you have use cases that we have thought about), please open a new issue.

@mgechev angular material and cdk are examples of libs which use global styles. If you use cdk in your lib and need to override something how would you do it without the global styles? Or If your lib uses material and wants to provider custom theme.

**Comment by JanneHarju** [comment]
> To overcome this situation :
> 
> * Use Angular-cli 7.1 which has a new option (--deleteOutputPath=false) and use the rimraf command to enrue dist cleaning is proceeded before any other script is executed.

It seems that this options is taken away from cli in Angular CLI: 7.3.1 version. I didn't find any information about it and didn't find any replacement for this options. So I started use wait-on package to wait until package-name.d.ts file is ready until to start scss-bundle.

And for others trying to find -w options in scss-bundle you should update your package to 2.5.0-beta. It is new feature of that package.

**Comment by JanneHarju** [comment]
> To overcome this situation :
> 
> * Use Angular-cli 7.1 which has a new option (--deleteOutputPath=false) and use the rimraf command to enrue dist cleaning is proceeded before any other script is executed.
> * If using previous Angular version : add a short delay before using cpx, scss-bundle (using npm package like @sindresorhus delay-cli)

I'm still wondering how your solution can even work with library because if I am understand correctly there is no --delete-output-path parameter with ng-packagr like normal angular application has. Check this out https://github.com/ng-packagr/ng-packagr/issues/632
So I'm stuck with your other suggestion.
